failure in tcp/test_ucp_peer_failure.zcopy/1

### Describe the bug
Happened here: http://hpc-master.lab.mtl.com:8080/blue/organizations/jenkins/ucx/detail/ucx/7119/pipeline/

```
[2020-09-16T09:40:00.862Z] [ RUN      ] tcp/test_ucp_peer_failure.zcopy/1 <tcp>
[2020-09-16T09:40:01.118Z] [     INFO ] < tcp_ep 0x1002d9abf90 (state=CONNECTED): send(31) failed: Connection reset by remote peer >
[2020-09-16T09:55:08.558Z] /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/test_helpers.cc:48: Failure
[2020-09-16T09:55:08.558Z] Failed
[2020-09-16T09:55:08.558Z] Connection timed out - abort testing
[2020-09-16T09:55:08.558Z] [r-vmb-ppc-jenkins:18870:0:18870] Caught signal 6 (Aborted: tkill(2) or tgkill(2))
[2020-09-16T09:55:08.558Z] 
[2020-09-16T09:55:08.558Z] /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/api/uct.h: [ uct_ep_am_zcopy() ]
[2020-09-16T09:55:08.558Z]       ...
[2020-09-16T09:55:08.558Z]      2665                                             uct_completion_t *comp)
[2020-09-16T09:55:08.558Z]      2666 {
[2020-09-16T09:55:08.558Z]      2667     return ep->iface->ops.ep_am_zcopy(ep, id, header, header_length, iov, iovcnt,
[2020-09-16T09:55:08.558Z] ==>  2668                                       flags, comp);
[2020-09-16T09:55:08.558Z]      2669 }
[2020-09-16T09:55:08.558Z]      2670 
[2020-09-16T09:55:08.558Z]      2671 /**
[2020-09-16T09:55:08.558Z] 
[2020-09-16T09:55:08.558Z] ==== backtrace (tid:  18870) ====
[2020-09-16T09:55:08.558Z]  0 0x000000000005eba8 ucs_debug_print_backtrace()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucs/debug/debug.c:656
[2020-09-16T09:55:08.558Z]  1 0x0000000000078884 uct_ep_am_zcopy()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/api/uct.h:2668
[2020-09-16T09:55:08.558Z]  2 0x0000000000078884 ucp_do_am_zcopy_multi()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/proto/proto_am.inl:380
[2020-09-16T09:55:08.558Z]  3 0x0000000000078884 ucp_tag_eager_zcopy_multi()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/tag/eager_snd.c:186
[2020-09-16T09:55:08.558Z]  4 0x000000000002592c uct_tcp_ep_pending_queue_dispatch()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_ep.c:497
[2020-09-16T09:55:08.558Z]  5 0x000000000002d978 uct_tcp_cm_change_conn_state()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_cm.c:49
[2020-09-16T09:55:08.558Z]  6 0x000000000002e3d4 uct_tcp_cm_handle_conn_ack()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_cm.c:544
[2020-09-16T09:55:08.558Z]  7 0x000000000002e610 uct_tcp_cm_handle_conn_pkt()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_cm.c:568
[2020-09-16T09:55:08.558Z]  8 0x0000000000027980 uct_tcp_ep_progress_am_rx()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_ep.c:1004
[2020-09-16T09:55:08.558Z]  9 0x0000000000027980 uct_tcp_ep_progress_data_rx()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_ep.c:1079
[2020-09-16T09:55:08.558Z] 10 0x000000000002b784 uct_tcp_iface_handle_events()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_iface.c:205
[2020-09-16T09:55:08.558Z] 11 0x000000000006fbcc ucs_event_set_wait()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucs/sys/event_set.c:213
[2020-09-16T09:55:08.558Z] 12 0x000000000002acec uct_tcp_iface_progress()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/tcp/tcp_iface.c:222
[2020-09-16T09:55:08.558Z] 13 0x0000000000040b34 ucs_callbackq_dispatch()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucs/datastruct/callbackq.h:211
[2020-09-16T09:55:08.558Z] 14 0x0000000000040b34 uct_worker_progress()  /scrap/jenkins/workspace/ucx-3/contrib/../src/uct/api/uct.h:2405
[2020-09-16T09:55:08.558Z] 15 0x0000000000040b34 ucp_worker_progress()  /scrap/jenkins/workspace/ucx-3/contrib/../src/ucp/core/ucp_worker.c:2191
[2020-09-16T09:55:08.558Z] 16 0x00000000105dbf5c ucp_test_base::entity::progress()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/ucp_test.cc:764
[2020-09-16T09:55:08.558Z] 17 0x00000000105dc060 ucp_test::progress()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/ucp_test.cc:147
[2020-09-16T09:55:08.558Z] 18 0x000000001041666c test_ucp_peer_failure::do_test()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/test_ucp_peer_failure.cc:341
[2020-09-16T09:55:08.558Z] 19 0x0000000010203708 ucs::test_base::run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/test.cc:311
[2020-09-16T09:55:08.558Z] 20 0x0000000010203708 ucs::test_base::TestBodyProxy()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/test.cc:337
[2020-09-16T09:55:08.558Z] 21 0x00000000103d0210 ucp_test::TestBody()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/ucp/ucp_test.h:174
[2020-09-16T09:55:08.558Z] 22 0x00000000101dc61c HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-09-16T09:55:08.558Z] 23 0x00000000101dc61c testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3598
[2020-09-16T09:55:08.558Z] 24 0x00000000101cb7ac testing::Test::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3635
[2020-09-16T09:55:08.558Z] 25 0x00000000101cb90c testing::TestInfo::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3812
[2020-09-16T09:55:08.558Z] 26 0x00000000101cbb58 testing::TestCase::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3930
[2020-09-16T09:55:08.558Z] 27 0x00000000101d1b58 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:5808
[2020-09-16T09:55:08.558Z] 28 0x00000000101d1fa4 testing::internal::UnitTestImpl::RunAllTests()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:5725
[2020-09-16T09:55:08.558Z] 29 0x00000000101d1fa4 HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3562
[2020-09-16T09:55:08.558Z] 30 0x00000000101d1fa4 HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:3598
[2020-09-16T09:55:08.558Z] 31 0x00000000101d1fa4 testing::UnitTest::Run()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest-all.cc:5422
[2020-09-16T09:55:08.558Z] 32 0x000000001014f328 RUN_ALL_TESTS()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/gtest.h:20059
[2020-09-16T09:55:08.558Z] 33 0x000000001014f328 main()  /scrap/jenkins/workspace/ucx-3/contrib/../test/gtest/common/main.cc:102
[2020-09-16T09:55:08.558Z] 34 0x0000000000025100 generic_start_main.isra.0()  libc-start.c:0
[2020-09-16T09:55:08.558Z] 35 0x00000000000252f4 __libc_start_main()  ???:0
[2020-09-16T09:55:08.558Z] =================================
[2020-09-16T09:55:08.558Z] Sending notification to mikhailb@nvidia.com
[2020-09-16T09:55:08.558Z] [r-vmb-ppc-jenkins:18870:0:18870] Process frozen...
[2020-09-16T11:58:21.699Z] make: *** [test] Terminated
```

--- a/src/ucp/proto/proto_am.inl
+++ b/src/ucp/proto/proto_am.inl
@@ -426,12 +426,17 @@ ucs_status_t ucp_do_am_zcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
                 ucs_assert(status == UCS_INPROGRESS);
                 return UCS_OK;
             }
+
+            return UCS_ERR_NO_RESOURCE;
         }
 
         ucp_request_send_state_advance(req, &state,
                                        UCP_REQUEST_SEND_PROTO_ZCOPY_AM,
                                        status);
         if (UCS_STATUS_IS_ERR(status)) {
+            if (req->send.state.uct_comp.count == 0) {
+               complete(req, status);
+            }
             return status;
         } else {
             if (enable_am_bw) {

--- a/src/uct/tcp/tcp_ep.c
+++ b/src/uct/tcp/tcp_ep.c
@@ -571,6 +571,37 @@ static void uct_tcp_ep_handle_disconnected(uct_tcp_ep_t *ep, ucs_status_t status
     }
 }
 
+static inline ucs_status_t uct_tcp_ep_handle_send_err(uct_tcp_ep_t *ep,
+                                                      ucs_status_t status)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+
+    ucs_assert(status != UCS_ERR_NO_PROGRESS);
+
+    status = uct_tcp_ep_handle_io_err(ep, "send", status);
+    if (status == UCS_ERR_CANCELED) {
+        /* If no data were read to the allocated buffer,
+         * we can safely reset it for futher re-use and to
+         * avoid overwriting this buffer, because `rx::length == 0` */
+        if (ep->tx.length == 0) {
+            uct_tcp_ep_ctx_reset(&ep->tx);
+        }
+    } else {
+        uct_tcp_ep_handle_disconnected(ep, status);
+        if (iface->super.err_handler != NULL) {
+            /* Translate the error to EP timeout error, since user expects that
+             * status returned from UCT error callback is the same as it is
+             * returned from the UCT EP send operation (or from UCT competion
+             * callback)
+             * TODO: revise this behavior */
+            return UCS_ERR_ENDPOINT_TIMEOUT;
+        }
+    }
+
+    return status;
+}
+
 static inline ssize_t uct_tcp_ep_send(uct_tcp_ep_t *ep)
 {
     size_t sent_length;
@@ -584,7 +615,7 @@ static inline ssize_t uct_tcp_ep_send(uct_tcp_ep_t *ep)
                                 &sent_length);
     if (ucs_unlikely((status != UCS_OK) &&
                      (status != UCS_ERR_NO_PROGRESS))) {
-        return status;
+        return uct_tcp_ep_handle_send_err(ep, status);
     }
 
     uct_tcp_ep_tx_completed(ep, sent_length);
@@ -610,6 +641,7 @@ static inline ssize_t uct_tcp_ep_sendv(uct_tcp_ep_t *ep)
             return 0;
         }
 
+        status = uct_tcp_ep_handle_send_err(ep, status);
         uct_tcp_ep_zcopy_completed(ep, ctx->comp, status);
         return status;
     }
@@ -746,24 +778,6 @@ static inline void uct_tcp_ep_handle_recv_err(uct_tcp_ep_t *ep,
     }
 }
 
-static inline void uct_tcp_ep_handle_send_err(uct_tcp_ep_t *ep,
-                                              ucs_status_t status)
-{
-    ucs_assert(status != UCS_ERR_NO_PROGRESS);
-
-    status = uct_tcp_ep_handle_io_err(ep, "send", status);
-    if (status == UCS_ERR_CANCELED) {
-        /* If no data were read to the allocated buffer,
-         * we can safely reset it for futher re-use and to
-         * avoid overwriting this buffer, because `rx::length == 0` */
-        if (ep->tx.length == 0) {
-            uct_tcp_ep_ctx_reset(&ep->tx);
-        }
-    } else {
-        uct_tcp_ep_handle_disconnected(ep, status);
-    }
-}
-
 static inline unsigned uct_tcp_ep_recv(uct_tcp_ep_t *ep, size_t recv_length)
 {
     uct_tcp_iface_t UCS_V_UNUSED *iface = ucs_derived_of(ep->super.super.iface,
@@ -813,7 +827,6 @@ static unsigned uct_tcp_ep_progress_data_tx(uct_tcp_ep_t *ep)
         offset = (!(ep->flags & UCT_TCP_EP_FLAG_ZCOPY_TX) ?
                   uct_tcp_ep_send(ep) : uct_tcp_ep_sendv(ep));
         if (ucs_unlikely(offset < 0)) {
-            uct_tcp_ep_handle_send_err(ep, (ucs_status_t)offset);
             return 1;
         }
 
@@ -1172,7 +1185,6 @@ uct_tcp_ep_am_send(uct_tcp_ep_t *ep, const uct_tcp_am_hdr_t *hdr)
 
     offset = uct_tcp_ep_send(ep);
     if (ucs_unlikely(offset < 0)) {
-        uct_tcp_ep_handle_send_err(ep, (ucs_status_t)offset);
         return (ucs_status_t)offset;
     }
 
@@ -1220,8 +1232,7 @@ uct_tcp_ep_am_sendv(uct_tcp_ep_t *ep, int short_sendv, uct_tcp_am_hdr_t *hdr,
 
     status = ucs_socket_sendv_nb(ep->fd, iov, iov_cnt, &sent_length);
     if (ucs_unlikely((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS))) {
-        uct_tcp_ep_handle_send_err(ep, status);
-        return status;
+        return uct_tcp_ep_handle_send_err(ep, status);
     }
 
     uct_tcp_ep_tx_completed(ep, sent_length);

